docs/proposals/partition-ring-multi-az-replication.md

```diff
@@ -0,0 +1,209 @@
+
+### Protocol Buffer Definitions
+
+```protobuf
+message PartitionRingDesc {
+    map<string, PartitionDesc> partitions = 1;
+}
+
+message PartitionDesc {
+    PartitionState state = 1;
+    repeated uint32 tokens = 2;
+    map<string, InstanceDesc> instances = 3;
+    int64 registered_timestamp = 4;
+}
+
+// Unchanged from current implementation
+message InstanceDesc {
+    string addr = 1;
+    int64 timestamp = 2;
+    InstanceState state = 3;
+    string zone = 7;
+    int64 registered_timestamp = 8;
+}
+```
+
+### Partition States
+
+Partitions maintain a simplified state model that provides **clear ownership** where each series belongs to exactly one partition, but requires **additional state management** for partition states and lifecycle management:
+
+```go
+type PartitionState int
+
+const (
+    NON_READY PartitionState = iota  // Insufficient ingesters
+    ACTIVE                           // Fully operational
+    READONLY                         // Scale-down in progress
+)
+```
+
+State transitions:
+```mermaid
+stateDiagram-v2
+    [*] --> NON_READY
+    NON_READY --> ACTIVE : Required ingesters joined<br/>across all AZs
+    ACTIVE --> READONLY : Scale-down initiated
+    ACTIVE --> NON_READY : Ingester removed
+    READONLY --> NON_READY : Ingesters removed
+    NON_READY --> [*] : Partition deleted
+```
+
+### Partition Lifecycle Management
+
+#### Creating Partitions
+
+When a new ingester joins the ring:
+1. Check if a suitable partition exists with available slots
+2. If no partition exists, create a new partition in `NON_READY` state
+3. Add partition's tokens to the ring
+4. Add the ingester to the partition
+5. Wait for required number of ingesters across all AZs (one per AZ)
+6. Once all AZs are represented, transition partition to `ACTIVE`
+
+#### Removing Partitions
+
+The scale-down process follows these steps:
+1. **Mark READONLY**: Partition stops accepting new writes but continues serving reads
+2. **Data Transfer**: Wait for all ingesters in partition to transfer data and become empty
+3. **Coordinated Removal**: Remove one ingester from each AZ simultaneously
+4. **State Transition**: Partition automatically transitions to `NON_READY` (insufficient replicas)
+5. **Cleanup**: Remove remaining ingesters and delete partition from ring
+
+If not using READONLY mode, removing an ingester will make the partition as NON_READY. When all ingesters are removed, the last will delete the partition if configuration `unregister_on_shutdown` is true
+
+### Multi-Ring Migration Strategy
+
+To address the migration challenge for production clusters currently running token-based rings, this proposal also introduces a multi-ring infrastructure that allows gradual traffic shifting from token-based to partition-based rings:
```

Comment:
How would this migration infrastructure look like in Cortex? Is it part of the Cortex repo or will be a separate migration tool

Reply:
It depends on each use case, but mainly manually via runtime config. I do not pretend to create a new migration tool in this proposal. 
If the ingesters can sustain resharding of the ring, it can be similar to ring migration where we deploy a multi ring and then remove the old one. It gets more complex if we need to do by waves using percentage.

Reply:
@danielblando The changes are pretty big since the ring implementation is completely different. I will see what I can do to open-source it. 

Your understanding is correct, it is purely for user isolation and it is not a shard.